packages/core/src/components/tabs2/tabs.tsx

```diff
@@ -0,0 +1,285 @@
+        if (selectedTabId !== undefined) {
+            // keep state in sync with controlled prop, so state is canonical source of truth
+            this.setState({ selectedTabId });
+        }
+    }
+
+    public componentDidUpdate(_: ITabs2Props, prevState: ITabs2State) {
+        if (this.state.selectedTabId !== prevState.selectedTabId) {
+            this.moveSelectionIndicator();
+        }
+    }
+
+    /** Filters this.props.children to only `<Tab>`s */
+    private getTabChildren() {
+        return React.Children.toArray(this.props.children).filter(isTab) as TabElement[];
+    }
+
+    /** Queries root HTML element for all `.pt-tab`s with optional filter selector */
+    private getTabElements(subselector = "") {
+        if (this.tabElement == null) {
+            return [] as Elements;
+        }
+        return this.tabElement.queryAll(TAB_SELECTOR + subselector);
+    }
+
+    private getTabClickHandler(newTabId: TabId) {
+        return () => {
+            const isControlled = this.props.selectedTabId !== undefined;
+            const { selectedTabId } = isControlled ? this.props : this.state;
+            if (newTabId !== selectedTabId) {
+                safeInvoke(this.props.onChange, newTabId, selectedTabId);
+                if (!isControlled) {
+                    this.setState({ selectedTabId: newTabId });
+                }
+            }
+        };
+    }
+
+    private handleKeyDown = (e: React.KeyboardEvent<HTMLDivElement>) => {
+        const focusedElement = document.activeElement.closest(TAB_SELECTOR);
+        // rest of this is potentially expensive and futile, so bail if no tab is focused
+        if (focusedElement == null) { return; }
+
+        // must rely on DOM state because we have no way of mapping `focusedElement` to a JSX.Element
+        const enabledTabElements = this.getTabElements()
+            .filter((el) => el.getAttribute("aria-disabled") === "false");
+        const focusedIndex = enabledTabElements.indexOf(focusedElement);
+
+        if (focusedIndex >= 0 && isEventKeyCode(e, Keys.ARROW_LEFT, Keys.ARROW_RIGHT)) {
+            // UP keycode is between LEFT and RIGHT so this produces 1 | -1
+            const direction = e.which - Keys.ARROW_UP;
+            const { length } = enabledTabElements;
+            // auto-wrapping at 0 and `length`
+            const nextFocusedIndex = (focusedIndex + direction + length) % length;
+            (enabledTabElements[nextFocusedIndex] as HTMLElement).focus();
+        }
+    }
+
+    private handleKeyPress = (e: React.KeyboardEvent<HTMLDivElement>) => {
+        const targetTabElement = (e.target as HTMLElement).closest(TAB_SELECTOR) as HTMLElement;
+        if (targetTabElement != null && isEventKeyCode(e, Keys.SPACE, Keys.ENTER)) {
+            e.preventDefault();
+            targetTabElement.click();
+        }
+    }
+
+    private handleTabRef = (tabElement: HTMLDivElement) => { this.tabElement = tabElement; };
```

Comment:
yep that's ok and expected. i check for existence before using `this.tabElement` elsewhere.